include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -0,0 +1,1265 @@
+    {
+        // if the point has already been sampled, return the stored operational status
+        if (const auto op_value = has_already_been_sampled(sp); op_value.has_value())
+        {
+            return *op_value;
+        }
+
+        // fetch the x and y dimension values
+        const auto param_point = to_parameter_point(sp);
+
+        const auto operational = [this, &param_point]()
+        {
+            op_domain.operational_values[param_point] = operational_domain::operational_status::OPERATIONAL;
+
+            return operational_domain::operational_status::OPERATIONAL;
+        };
+
+        const auto non_operational = [this, &param_point]()
+        {
+            op_domain.operational_values[param_point] = operational_domain::operational_status::NON_OPERATIONAL;
+
+            return operational_domain::operational_status::NON_OPERATIONAL;
+        };
+
+        // increment the number of evaluated parameter combinations
+        ++num_evaluated_parameter_combinations;
+
+        // take the first (and only) output BDL pair
+        const auto& output_bdl_pair = output_bdl_pairs.front();
+
+        // initialize a BDL input iterator
+        bdl_input_iterator<Lyt> bii{layout, params.bdl_params};
+
+        sidb_simulation_parameters sim_params = params.sim_params;
+        set_x_dimension_value(sim_params, param_point.x);
+        set_y_dimension_value(sim_params, param_point.y);
+
+        // for each input combination
+        for (auto i = 0u; i < truth_table.num_bits(); ++i, ++bii)
+        {
+            // the expected output of the layout is the i-th bit of the truth table
+            const auto expected_output = kitty::get_bit(truth_table, i);
+
+            ++num_simulator_invocations;
+
+            if (can_positive_charges_occur(*bii, sim_params))
+            {
+                return non_operational();
+            }
+
+            sidb_simulation_result<Lyt> sim_result{};
+
+            if (params.sim_engine == sidb_simulation_engine::EXGS)
+            {
+                // perform an exhaustive ground state simulation
+                sim_result = exhaustive_ground_state_simulation(*bii, sim_params);
+            }
+            else if (params.sim_engine == sidb_simulation_engine::QUICKSIM)
+            {
+                // perform a heuristic simulation
+                const quicksim_params qs_params{sim_params, 500, 0.6, 1};
+                sim_result = quicksim(*bii, qs_params);
+            }
+            else
+            {
+                assert(false && "unsupported simulation engine");
+            }
+
+            // if no physically-valid charge distributions were found, the layout is non-operational
+            if (sim_result.charge_distributions.empty())
+            {
+                return non_operational();
+            }
+
+            // find the ground state, which is the charge distribution with the lowest energy
+            const auto ground_state = std::min_element(
```

Comment:
If we have a layout with a degenerated ground state, this could cause some problems in rare cases (the computation could be even hardware dependent).

Reply:
What would be your suggestion?

Reply:
I think there are a few ways to do this. Maybe one way to do it:
```suggestion
            // ground state is degenerate
            if (energy_distribution(sim_result.charge_distributions)[0]>1)
            {
                return non_operational();
            }
```

Reply:
Ok, I could do that. However, I still need the ground state to work with in the following lines.